generators/beagle/src/main/scala/BeaglePeriphery.scala

```diff
@@ -192,42 +265,141 @@ trait HasPeripheryBeagleBundle {
+trait HasPeripheryBeagleModuleImp extends LazyModuleImp with HasPeripheryBeagleBundle
+{
   val outer: HasPeripheryBeagle
 
+  // instantiate IOs
   val rst_async = IO(Input(Bool()))
   val boot = IO(Input(Bool()))
 
   val hbwif_clks = IO(Input(Vec(p(HbwifTLKey).numBanks, Clock()))) // clks for hbwif
   val hbwif_tx = IO(chiselTypeOf(outer.hbwif.module.tx))
   val hbwif_rx = IO(chiselTypeOf(outer.hbwif.module.rx))
 
-  // setup lbwif
   val lbwif_serial = IO(chiselTypeOf(outer.lbwif.module.io.ser))
-  lbwif_serial <> outer.lbwif.module.io.ser
 
-  val lbwifClkDiv = Module(new testchipip.ClockDivider(outer.scrParams.lbwifDividerBits))
-  val lbwif_clk = lbwifClkDiv.io.clockOut
-  val lbwif_rst = ResetCatchAndSync(lbwif_clk, reset.toBool)
-  lbwifClkDiv.io.divisor := outer.scr.module.io.lbwif_divider
+  // clocks coming from offchip
+  val single_clks = IO(Input(Vec(1, Clock())))
+  val diff_clks   = IO(Input(Vec(1, Clock())))
+
+  val bh_clk_sel     = IO(Input(Bool()))
+  val rs_clk_sel     = IO(Input(Bool()))
+  val uncore_clk_sel = IO(Input(Bool()))
+
+  val uncore_clk_out = IO(Output(Clock()))
+  val bh_clk_out     = IO(Output(Clock()))
+  val rs_clk_out     = IO(Output(Clock()))
+  val lbwif_clk_out  = IO(Output(Clock()))
+
+
+  // ------------------------------------------------
+
+  val scr_mod = outer.scr.module
+
+  // other signals
+  scr_mod.io.boot := boot
+  scr_mod.io.rst_async := rst_async
+
+  // ------------------------------------------------
+
+  // get the actual clock from the io clocks
+  require(bh_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+  require(rs_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+  require(uncore_clk_sel.getWidth >= (log2Ceil(single_clks.length) + log2Ceil(diff_clks.length)), "[sys-top] must be able to select all input clocks")
+
+  // setup uncore muxes and divider
+
+  val uncore_clk_mux = testchipip.ClockMutexMux(single_clks ++ diff_clks)
+  uncore_clk_mux.io.sel := uncore_clk_sel
+  uncore_clk_mux.io.resetAsync := rst_async
+  val uncore_clk_mux_out = uncore_clk_mux.io.clockOut
+  val uncore_clk_mux_rst = ResetCatchAndSync(uncore_clk_mux.io.clockOut, reset.asBool)
+
+  val uncore_clk_divider = withClockAndReset(uncore_clk_mux_out, uncore_clk_mux_rst) {
+    Module(new testchipip.ClockDivider(outer.scrParams.uncoreClkDivBits))
+  }
+  uncore_clk_divider.io.divisor := scr_mod.io.uncore_clk_divisor
+
+  // mux between the divided uncore and the undivided
+  val uncore_clks_pre_muxed = Seq(uncore_clk_mux_out, uncore_clk_divider.io.clockOut)
+  require(scr_mod.io.uncore_clk_pass_sel.getWidth >= log2Ceil(uncore_clks_pre_muxed.length), "[sys-top] must be able to select the uncore clocks")
+  val uncore_clk_pass_mux = testchipip.ClockMutexMux(uncore_clks_pre_muxed)
+  uncore_clk_pass_mux.io.sel := scr_mod.io.uncore_clk_pass_sel
```

Comment:
Just make sure all these select signals come up as passthrough.

Reply:
All the select signals for the muxes come from offchip except for the passthrough mux. That signal is given by a `AsyncResetReg` in the MMIO section, so it should come up correctly?